src/sage/modules/free_module_integer.py

```diff
@@ -804,3 +804,41 @@ def CVPP_2V(t, V, voronoi_cell):
             t_new = t_new - CVPP_2V(t_new, V_scaled, ZZ(2 ** (i - 1)) * voronoi_cell)
             i -= 1
         return t - t_new
+
+    def approximate_closest_vector(self, t):
```

Comment:
(As suggested in your "ideas for improvement") I think `delta` should be a user-specified argument. You might also want to add a note that this method changes the lattice basis (`self.LLL()` seems to modify `self`).

Reply:
In fact, why not just add `*args, **kwds` that is passed on to the `LLL` call? That will include e.g. `delta` and `algorithm`. Though that will probably complicate the bound in the docstring below.

Reply:
We don't want to reduce the lattice every time we call this method, so the way I envision it is that instead of checking `_basis_is_LLL_reduced` it keeps track of the delta value the basis has been reduced with. That becomes a little problematic if the basis has been reduced with BKZ or HKZ since we don't have an analog to delta (i think?). How do you suggest we deal with that?

Reply:
HKZ can reasonably be defined as using `delta=1` (though the given bound will likely be unnecessarily strict) or even a delta >1. BKZ does take a delta parameter but I have no experience using it, the block size should also affect the reduced basis and I'm not sure how to handle that.

Reply:
I'm leaning towards only reducing if the basis hasn't already been reduced, similar to how it is now, but we also allow the user to pass on arguments to LLL. These will be ignored if the basis is already LLL reduced.

Alternatively we only reduce if the user specifies reduction params.

*Alternatively* we simply leave it to the user to handle, potentially raising a warning if the basis hasn't been reduced at all.

Reply:
What about the user can call the function with any `*args, **kwds` and with an `algorithm="*"` argument. If the lattice basis hasn't been reduced by any algorithm yet (so there should be a `.is_reduced` flag?), then `algorithm` argument is required. Otherwise, it's optional

Reply:
There are cases when an already reduced basis is passed to `IntegerLattice`, I think this might introduce unnecessary complexity.

I now realize this should be the perfect use-case for `is_LLL_reduced`. It should run fairly quickly and accepts a `delta`-parameter. Then we could have the function accept a `delta` which has some default value (3/4 or 0.99), it checks if the basis is reduced to that degree and otherwise it runs LLL to make sure that it is. Sound good?

Reply:
Yes that works fine, I suggested mine since you wanted to handle BKZ too but I suppose it's output will also be LLL-reduced (I think)?

Reply:
I think it is a fairly universal measure of basis "goodness"